src/lib.rs

```diff
@@ -256,6 +259,28 @@ where
   _p: PhantomData<(C1, C2)>,
 }
 
+/// Final randomized fold of `RecursiveSNARK`
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct RandomLayerRecursive<E1, E2>
+where
+  E1: Engine<Base = <E2 as Engine>::Scalar>,
+  E2: Engine<Base = <E1 as Engine>::Scalar>,
+{
+  last_i: usize,
+  l_ur_primary: RelaxedR1CSInstance<E1>,
+  l_ur_secondary: RelaxedR1CSInstance<E2>,
+  nifs_Uf_secondary: NIFS<E2>,
+  nifs_Un_primary: NIFS<E1>,
+  nifs_Un_secondary: NIFS<E2>,
+  r_Wn_primary: RelaxedR1CSWitness<E1>,
+  r_Wn_secondary: RelaxedR1CSWitness<E2>,
+  // needed for CompressedSNARK proving,
```

Comment:
I'm confused a bit by this comment. These instances are required to verify the correctness of this type as well---regardless of compression right?

Reply:
If you were just verifying a randomized RecursiveSNARK, you would "refold" to get these instances. see [here](https://github.com/microsoft/Nova/blob/2a5cee6f222c651511faa4ddbc4746de57c4e0d2/src/lib.rs#L979) and [here](https://github.com/microsoft/Nova/blob/2a5cee6f222c651511faa4ddbc4746de57c4e0d2/src/lib.rs#L987)